drivers: random: Add nRF5 random number generator driver

durub opened the pull request:
Origin: Original

Signed-off-by: Thiago Silveira <thiago@exati.com.br>

Commit:
drivers: entropy: Add nRF5 entropy generator driver

Origin: Original

Signed-off-by: Thiago Silveira <thiago@exati.com.br>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

durub commented:
My first contribution, I hope everything is ok.

carlescufi commented:
@anangl could you take a look at this?

anangl commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,67 @@
+	nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
+
+	/* Start the nRF5 RNG generator */
+	nrf_rng_task_trigger(NRF_RNG_TASK_START);
```
This will make the RNG peripheral running (and consuming power) all the time, regardless of whether random values are needed or not. Consider starting and stopping the peripheral in `random_nrf5_get_entropy`. This will introduce some extra delay before the first random byte is generated, but it does not seem to be a problem.

durub replied:
Hi Andrzej,

That's a good point. I was optimizing for performance, but seeing as generating random numbers is fairly rare in a program and nRF52 family is supposed to be low power, I think I was optimizing for the wrong thing. This, and the fact that I didn't want to deal with concurrency... :stuck_out_tongue: 

I implemented your suggestion. Could you take another look?

To make it work, I implemented two mechanisms:
- An atomic user counter, which should solve higher priority threads/interrupts being scheduled, generating random numbers, and then turning off the peripheral, leaving the lower priority thread helpless;
- A timeout, which solves data races regarding turning off the peripheral. After all, we are performing a non-atomic operation after checking for atomic zero. However, I think these should be pretty rare if not impossible, considering RTOSes' scheduling.

durub commented:
I just implemented @anangl's suggestions.

@anangl, @cvinayak and @carlescufi, if possible, please review the new version!

Thanks a lot,
Thiago

tbursztyka commented on drivers/random/Kconfig.nrf5:
```diff
@@ -0,0 +1,34 @@
+	    nRF52832 - 30 uS
+	    nRF52840 - 30 uS
+
+menuconfig RANDOM_NRF5_BIAS_CORRECTION
```
s/menuconfig/config

tbursztyka commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,109 @@
+};
+
+DEVICE_AND_API_INIT(random_nrf5, CONFIG_RANDOM_NAME,
+			random_nrf5_init, &random_nrf5_data, NULL,
```
is it github's viewer or are the parameters not properly aligned to previous '(' + 1 space?
(i.e.: random_nrf5_init shoul be right below random_nrf5)

durub replied:
The parameters were not properly aligned before, my tab size was 4 instead of 8. I aligned it properly now, but github's viewer is acting up. They are properly aligned in the editor.

[aligned](https://i.imgur.com/pVaU3Ap.png)

tbursztyka reviewed (changes requested):
tiny comments

durub commented:
Hi @tbursztyka, thanks for the review. I fixed the points you mentioned. Could you take another look, please?

anangl commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,109 @@
+	/* Only stop the RNG generator peripheral if we're the last user */
+	if (atomic_get(&DEV_DATA(device)->user_count) == 0) {
+		nrf_rng_task_trigger(NRF_RNG_TASK_STOP);
+	}
```
What about using the VALRDY_STOP shortcut to turn off the peripheral, and getting rid of this timeout checking? I mean something like this:

```
atomic_inc(&DEV_DATA(device)->user_count);

nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
nrf_rng_task_trigger(NRF_RNG_TASK_START);

while (len) {
[..]
}

if (atomic_dec(&DEV_DATA(device)->user_count) == 1) {
	nrf_rng_shorts_enable(NRF_RNG_SHORT_VALRDY_STOP_MASK);

	if (atomic_get(&DEV_DATA(device)->user_count) != 0) {
		nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
		nrf_rng_task_trigger(NRF_RNG_TASK_START);
	}
}
```
The RNG peripheral will then be stopped after generating another random byte, unless some new request appears in the meantime.

This second `if` should solve the race condition when some higher priority code turns on the RNG peripheral and starts waiting for the VALRDY event right between `atomic_dec` and `nrf_rng_shorts_enable`.

Well, concurrency just complicates simple things... ;-)

durub replied:
Concurrency: it is what it is, for better or worse. At least we can say we work doing hard things... :smile: 

I like this solution. I'll implement and test it as soon as I have some time. Thanks a lot, Andrzej.

durub commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,109 @@
+
+	while (!nrf_rng_event_get(NRF_RNG_EVENT_VALRDY)) {
+		timeout++;
+		if (timeout >= 1000000) {
```
Ooops! That's what you get when you don't test _every_ case. Good catch!

durub commented:
Hi @anangl, thanks again for your time. I implemented your suggestion and tested every corner case imaginable (by me), could you take another look, please?

I decided to keep the timeout in case someone disabled the peripheral or did some funky stuff outside of the driver.

anangl commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,116 @@
+			 * Enable the peripheral again
+			 */
+			nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
+			nrf_rng_task_trigger(NRF_RNG_TASK_STOP);
```
`NRF_RNG_TASK_START`

durub replied:
Wow -- sorry about that and thanks for the catch. Sometimes it's best to go to sleep instead of programming, but that's the time I have to contribute. Fixed

anangl commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,116 @@
+
+static struct random_nrf5_dev_data random_nrf5_data;
+
+static struct random_driver_api random_nrf5_api_funcs = {
```
This could be `const`.

durub replied:
Done!

carlescufi commented on drivers/random/Kconfig.nrf5:
```diff
@@ -0,0 +1,34 @@
+
+menuconfig RANDOM_NRF5_RNG
+	bool "nRF5 RNG driver"
+	depends on RANDOM_GENERATOR
```
@cvinayak  would it be an issue if the 2 implementations coexisted?

cvinayak replied:
Add depends on !CONFIG_BT, if CONFIG_BT=y, then there is bt_rand public API in include/bluetooth/crypto.h

durub replied:
Ok -- I do believe that, in the future, the Bluetooth subsystem should move on to use the controller's device drivers, but that is not for this PR to fix.

When `CONFIG_BT=y`, should I just disable the RNG driver? Should the user then use `#ifdef`s to handle this case themselves, changing the function they call to get random data? How can we make this more user friendly?

Maybe just disabling the RNG driver when `CONFIG_BT=y` would be fine in a transitory period... but I think this needs a solution in the long term.

durub replied:
For now, let's just disable the driver when `CONFIG_BT=y` and handle these issues in future PRs. What do you think?

cvinayak replied:
yes, disable driver on CONFIG_BT=y. This is better than an unstable system when this driver and bt_rand co-exist.

durub replied:
@cvinayak, I disabled the driver when Bluetooth is enabled and tested the interaction on menuconfig. It's working properly. I think it's ready for merge now.

Thanks,
Thiago

cvinayak commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,61 @@
+static int random_nrf5_init(struct device *device)
+{
+    /* Enable or disable bias correction */
+    if (CONFIG_RANDOM_NRF5_BIAS_CORRECTION) {
```
IS_ENABLED(CONFIG_...)

cvinayak commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,61 @@
+u32_t sys_rand32_get(void)
+{
+    u32_t value;
+    random_nrf5_get_entropy(DEVICE_GET(random_nrf5), (u8_t *) &value, sizeof(value));
```
Missing blank line after variable declaration.

cvinayak commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,61 @@
+};
+
+DEVICE_AND_API_INIT(random_nrf5, CONFIG_RANDOM_NAME,
+            random_nrf5_init, NULL, NULL,
```
Indentation, tabs then spaces.  seems your editor is using spaces.

cvinayak commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,61 @@
+static inline u8_t random_nrf5_get_u8(void)
+{
+    while (!nrf_rng_event_get(NRF_RNG_EVENT_VALRDY));
+    u8_t value = nrf_rng_random_value_get();
```
blank lines after each line.

durub commented:
Wow, an old version of the code got in... seems some mishap from my side. I switched from the computer to the laptop (I am in a business trip) and probably forgot to update something.

@cvinayak

durub commented:
Fixed, sorry about that. @cvinayak

cvinayak commented on drivers/random/Kconfig.nrf5:
```diff
@@ -0,0 +1,35 @@
+menuconfig RANDOM_NRF5_RNG
+	bool "nRF5 RNG driver"
+	depends on RANDOM_GENERATOR && !BT
+	default n
```
No need for "default n", implicitly the default is n.

cvinayak commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,116 @@
+static int random_nrf5_init(struct device *device)
+{
+	/* Enable or disable bias correction */
+#ifdef CONFIG_RANDOM_NRF5_BIAS_CORRECTION
```
use if-then-else clause instead.

if (IS_ENABLED(CONFIG_...)) {
} else {
}

cvinayak commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,116 @@
+	u8_t value;
+	u32_t timeout = 0;
+
+	while (!nrf_rng_event_get(NRF_RNG_EVENT_VALRDY)) {
```
This loop is not low power, and use system clock interface to measure timeout (nRF51 is 16MHz CM0 CPU and nRF52 is 64MHz CM4F CPU).

durub replied:
I understand that the loop could be low power if I used the __WFE, __SVE, __WFE combo, but can I still have a timeout using these constructs? I don't understand the exact mechanics of it. Apart from "one-time" events like VALRDY, does the system generate events at fixed intervals, letting the loop continue?

When you say system clock interface, do you mean `k_cycle_get_32()` and related functions or something else?

cvinayak commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,116 @@
+static inline s16_t random_nrf5_get_u8(void)
+{
+	u8_t value;
+	u32_t timeout = 0;
```
Reverse tree declaration. longer lines first.

cvinayak commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	value = nrf_rng_random_value_get();
+	nrf_rng_event_clear(NRF_RNG_EVENT_VALRDY);
```
Blank line between unrelated statements.

durub commented:
@cvinayak, I addressed your concerns in the latest update.

I ended up removing the timeout, as I don't think it is that important now that we don't use the driver when CONFIG_BT is enabled.

anangl commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,115 @@
+	u8_t value;
+
+	while (!nrf_rng_event_get(NRF_RNG_EVENT_VALRDY)) {
+		__WFE();
```
This is not going to work properly. The VALRDY event itself won't wake up the core. You would need to generate an interrupt for this to happen (see for instance [here](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0553a/CHDEAAHJ.html)). VALRDY is a peripheral event, not to be confused with the internal event register in the core that influences WFE.

lpereira commented on drivers/random/random_nrf5.c:
```diff
@@ -0,0 +1,115 @@
+		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		    &random_nrf5_api_funcs);
+
+u32_t sys_rand32_get(void)
```
This function can be removed as of today -- the `sys_rand32_get()` function is now driver-independent.

carlescufi commented:
@lpereira thanks for the info.
@durub We would really like to fix and merge this. If you are not available we will try to push to your branch to fix our review comments. If you want to do this yourself please let us know.

cvinayak commented on drivers/random/Kconfig.nrf5:
```diff
@@ -0,0 +1,33 @@
+# Kconfig.nrf5 - nRF5 random number generator driver configuration
+#
+# Copyright (c) 2017 Exati Tecnologia Ltda.
```
Copyright Nordic semi

cvinayak commented on drivers/random/Kconfig.nrf5:
```diff
@@ -0,0 +1,33 @@
+	  generator, based on internal thermal noise, that provides a
+	  random 8-bit value to the host when read.
+
+	  Run time per byte is as follows:
```
These number not needed here, product specification document has them.

durub commented:
Hi everyone,

Sorry for the lack of an answer, I was on vacation for one month and only came back this week.
@carlescufi, @anangl and @cvinayak, thanks for fixing the PR and getting it merged in.

Cheers!
